File: clients/python/retab/resources/documents/extractions.py
import base64
import json
from io import IOBase
from pathlib import Path
from typing import Any, AsyncGenerator, Generator

from anthropic.types.message_param import MessageParam
from openai.types.chat import ChatCompletionMessageParam
from openai.types.chat.chat_completion_reasoning_effort import ChatCompletionReasoningEffort
from openai.types.chat.parsed_chat_completion import ParsedChatCompletionMessage
from openai.types.responses.response import Response
from openai.types.responses.response_input_param import ResponseInputItemParam
from pydantic_core import PydanticUndefined
from pydantic import HttpUrl

from ..._resource import AsyncAPIResource, SyncAPIResource
from ...utils.ai_models import assert_valid_model_extraction
from ...utils.json_schema import filter_auxiliary_fields_json, load_json_schema, unflatten_dict
from ...utils.mime import MIMEData, prepare_mime_document
from ...utils.stream_context_managers import as_async_context_manager, as_context_manager
from ...types.chat import ChatCompletionRetabMessage
from ...types.documents.extractions import DocumentExtractRequest, LogExtractionRequest, RetabParsedChatCompletion, RetabParsedChatCompletionChunk, RetabParsedChoice
from ...types.browser_canvas import BrowserCanvas
from ...types.modalities import Modality
from ...types.schemas.object import Schema
from ...types.standards import PreparedRequest


def maybe_parse_to_pydantic(schema: Schema, response: RetabParsedChatCompletion, allow_partial: bool = False) -> RetabParsedChatCompletion:
    if response.choices[0].message.content:
        try:
            if allow_partial:
                response.choices[0].message.parsed = schema._partial_pydantic_model.model_validate(filter_auxiliary_fields_json(response.choices[0].message.content))
            else:
                response.choices[0].message.parsed = schema.pydantic_model.model_validate(filter_auxiliary_fields_json(response.choices[0].message.content))
        except Exception:
            pass
    return response


class BaseExtractionsMixin:
    def prepare_extraction(
        self,
        json_schema: dict[str, Any] | Path | str,
        document: Path | str | IOBase | HttpUrl | None = None,
        documents: list[Path | str | IOBase | HttpUrl] | None = None,
        image_resolution_dpi: int = PydanticUndefined,  # type: ignore[assignment]
        browser_canvas: BrowserCanvas = PydanticUndefined,  # type: ignore[assignment]
        model: str = PydanticUndefined,  # type: ignore[assignment]
        temperature: float = PydanticUndefined,  # type: ignore[assignment]
        modality: Modality = PydanticUndefined,  # type: ignore[assignment]
        reasoning_effort: ChatCompletionReasoningEffort = PydanticUndefined,  # type: ignore[assignment]
        stream: bool = False,
        n_consensus: int = PydanticUndefined,  # type: ignore[assignment]
        store: bool = False,
        idempotency_key: str | None = None,
    ) -> PreparedRequest:
        assert_valid_model_extraction(model)

        json_schema = load_json_schema(json_schema)

        # Handle both single document and multiple documents
        if document is not None and documents is not None:
            raise ValueError("Cannot provide both 'document' and 'documents' parameters. Use either one.")

        # Convert single document to documents list for consistency
        if document is not None:
            processed_documents = [prepare_mime_document(document)]
        elif documents is not None:
            processed_documents = [prepare_mime_document(doc) for doc in documents]
        else:
            raise ValueError("Must provide either 'document' or 'documents' parameter.")

        # Validate DocumentAPIRequest data (raises exception if invalid)
        request = DocumentExtractRequest(
            json_schema=json_schema,
            documents=processed_documents,
            model=model,
            temperature=temperature,
            stream=stream,
            modality=modality,
            store=store,
            reasoning_effort=reasoning_effort,
            n_consensus=n_consensus,
            image_resolution_dpi=image_resolution_dpi,
            browser_canvas=browser_canvas,
        )

        return PreparedRequest(
            method="POST", url="/v1/documents/extractions", data=request.model_dump(mode="json", exclude_unset=True, exclude_defaults=True), idempotency_key=idempotency_key
        )

    def prepare_log_extraction(
        self,
        document: Path | str | IOBase | HttpUrl | None,
        json_schema: dict[str, Any],
        model: str,
        temperature: float,
        completion: Any | None = None,
        # The messages can be provided in different formats, we will convert them to the Retab-compatible format
        messages: list[ChatCompletionRetabMessage] | None = None,
        openai_messages: list[ChatCompletionMessageParam] | None = None,
        anthropic_messages: list[MessageParam] | None = None,
        anthropic_system_prompt: str | None = None,
        # New fields for the Responses API
        openai_responses_input: list[ResponseInputItemParam] | None = None,
        openai_responses_output: Response | None = None,
    ) -> PreparedRequest:
        if document is None:
            mime_document = MIMEData(
                filename="dummy.txt",
                # url is a base64 encoded string with the mime type and the content. For the dummy one we will send a .txt file with the text "No document provided"
                url="data:text/plain;base64," + base64.b64encode(b"No document provided").decode("utf-8"),
            )
        else:
            mime_document = prepare_mime_document(document)

        return PreparedRequest(
            method="POST",
            url="/v1/documents/log_extraction",
            data=LogExtractionRequest(
                document=mime_document,
                messages=messages,
                openai_messages=openai_messages,
                anthropic_messages=anthropic_messages,
                anthropic_system_prompt=anthropic_system_prompt,
                completion=completion,
                openai_responses_input=openai_responses_input,
                openai_responses_output=openai_responses_output,
                json_schema=json_schema,
                model=model,
                temperature=temperature,
            ).model_dump(mode="json"),
            raise_for_status=True,
        )


class Extractions(SyncAPIResource, BaseExtractionsMixin):
    """Extraction API wrapper"""

    def parse(
        self,
        json_schema: dict[str, Any] | Path | str,
        model: str,
        document: Path | str | IOBase | HttpUrl | None = None,
        documents: list[Path | str | IOBase | HttpUrl] | None = None,
        image_resolution_dpi: int = PydanticUndefined,  # type: ignore[assignment]
        browser_canvas: BrowserCanvas = PydanticUndefined,  # type: ignore[assignment]
        temperature: float = PydanticUndefined,  # type: ignore[assignment]
        modality: Modality = PydanticUndefined,  # type: ignore[assignment]
        reasoning_effort: ChatCompletionReasoningEffort = PydanticUndefined,  # type: ignore[assignment]
        n_consensus: int = PydanticUndefined,  # type: ignore[assignment]
        idempotency_key: str | None = None,
        store: bool = False,
    ) -> RetabParsedChatCompletion:
        """
        Process one or more documents using the Retab API.

        Args:
            json_schema: JSON schema defining the expected data structure
            model: The AI model to use for processing
            document: Single document to process (use either this or documents, not both)
            documents: List of documents to process (use either this or document, not both)
            image_resolution_dpi: Optional image resolution DPI
            browser_canvas: Optional browser canvas size
            temperature: Model temperature setting (0-1)
            modality: Modality of the document (e.g., native)
            reasoning_effort: The effort level for the model to reason about the input data.
            n_consensus: Number of consensus extractions to perform (default: 1 which computes a single extraction and the likelihoods comes from the model logprobs)
            idempotency_key: Idempotency key for request
            store: Whether to store the document in the Retab database
        Returns:
            RetabParsedChatCompletion: Parsed response from the API
        Raises:
            ValueError: If neither document nor documents is provided, or if both are provided
            HTTPException: If the request fails
        """

        # Validate DocumentAPIRequest data (raises exception if invalid)
        request = self.prepare_extraction(
            json_schema=json_schema,
            document=document,
            documents=documents,
            image_resolution_dpi=image_resolution_dpi,
            browser_canvas=browser_canvas,
            model=model,
            temperature=temperature,
            modality=modality,
            reasoning_effort=reasoning_effort,
            stream=False,
            n_consensus=n_consensus,
            store=store,
            idempotency_key=idempotency_key,
        )
        response = self._client._prepared_request(request)

        schema = Schema(json_schema=load_json_schema(json_schema))
        return maybe_parse_to_pydantic(schema, RetabParsedChatCompletion.model_validate(response))

    @as_context_manager
    def stream(
        self,
        json_schema: dict[str, Any] | Path | str,
        model: str,
        document: Path | str | IOBase | HttpUrl | None = None,
        documents: list[Path | str | IOBase | HttpUrl] | None = None,
        image_resolution_dpi: int = PydanticUndefined,  # type: ignore[assignment]
        browser_canvas: BrowserCanvas = PydanticUndefined,  # type: ignore[assignment]
        temperature: float = PydanticUndefined,  # type: ignore[assignment]
        modality: Modality = PydanticUndefined,  # type: ignore[assignment]
        reasoning_effort: ChatCompletionReasoningEffort = PydanticUndefined,  # type: ignore[assignment]
        n_consensus: int = PydanticUndefined,  # type: ignore[assignment]
        idempotency_key: str | None = None,
        store: bool = False,
    ) -> Generator[RetabParsedChatCompletion, None, None]:
        """
        Process one or more documents using the Retab API with streaming enabled.

        Args:
            json_schema: JSON schema defining the expected data structure
            model: The AI model to use for processing
            document: Single document to process (use either this or documents, not both)
            documents: List of documents to process (use either this or document, not both)
            image_resolution_dpi: Optional image resolution DPI.
            browser_canvas: Optional browser canvas size.
            temperature: Model temperature setting (0-1)
            modality: Modality of the document (e.g., native)
            reasoning_effort: The effort level for the model to reason about the input data.
            n_consensus: Number of consensus extractions to perform (default: 1 which computes a single extraction and the likelihoods comes from the model logprobs)
            idempotency_key: Idempotency key for request
            store: Whether to store the document in the Retab database

        Returns:
            Generator[RetabParsedChatCompletion]: Stream of parsed responses
        Raises:
            ValueError: If neither document nor documents is provided, or if both are provided
            HTTPException: If the request fails
        Usage:
        ```python
        # Single document
        with retab.documents.extractions.stream(json_schema, model, document=document) as stream:
            for response in stream:
                print(response)

        # Multiple documents
        with retab.documents.extractions.stream(json_schema, model, documents=[doc1, doc2]) as stream:
            for response in stream:
                print(response)
        ```
        """
        request = self.prepare_extraction(
            json_schema=json_schema,
            document=document,
            documents=documents,
            image_resolution_dpi=image_resolution_dpi,
            browser_canvas=browser_canvas,
            model=model,
            temperature=temperature,
            modality=modality,
            reasoning_effort=reasoning_effort,
            stream=True,
            n_consensus=n_consensus,
            store=store,
            idempotency_key=idempotency_key,
        )
        schema = Schema(json_schema=load_json_schema(json_schema))

        # Request the stream and return a context manager
        ui_parsed_chat_completion_cum_chunk: RetabParsedChatCompletionChunk | None = None
        # Initialize the RetabParsedChatCompletion object
        ui_parsed_completion: RetabParsedChatCompletion = RetabParsedChatCompletion(
            id="",
            created=0,
            model="",
            object="chat.completion",
            likelihoods={},
            choices=[
                RetabParsedChoice(
                    index=0,
                    message=ParsedChatCompletionMessage(content="", role="assistant"),
                    finish_reason=None,
                    logprobs=None,
                )
            ],
        )
        for chunk_json in self._client._prepared_request_stream(request):
            if not chunk_json:
                continue
            ui_parsed_chat_completion_cum_chunk = RetabParsedChatCompletionChunk.model_validate(chunk_json).chunk_accumulator(ui_parsed_chat_completion_cum_chunk)
            # Basic stuff
            ui_parsed_completion.id = ui_parsed_chat_completion_cum_chunk.id
            ui_parsed_completion.created = ui_parsed_chat_completion_cum_chunk.created
            ui_parsed_completion.model = ui_parsed_chat_completion_cum_chunk.model
            # Update the ui_parsed_completion object
            parsed = unflatten_dict(ui_parsed_chat_completion_cum_chunk.choices[0].delta.flat_parsed)
            likelihoods = unflatten_dict(ui_parsed_chat_completion_cum_chunk.choices[0].delta.flat_likelihoods)
            ui_parsed_completion.choices[0].message.content = json.dumps(parsed)
            ui_parsed_completion.choices[0].message.parsed = parsed
            ui_parsed_completion.likelihoods = likelihoods

            yield maybe_parse_to_pydantic(schema, ui_parsed_completion, allow_partial=True)

        # change the finish_reason to stop
        ui_parsed_completion.choices[0].finish_reason = "stop"
        yield maybe_parse_to_pydantic(schema, ui_parsed_completion)

    def log(
        self,
        document: Path | str | IOBase | HttpUrl | None,
        json_schema: dict[str, Any],
        model: str,
        temperature: float,
        completion: Any | None = None,
        # The messages can be provided in different formats, we will convert them to the Retab-compatible format
        messages: list[ChatCompletionRetabMessage] | None = None,
        openai_messages: list[ChatCompletionMessageParam] | None = None,
        anthropic_messages: list[MessageParam] | None = None,
        anthropic_system_prompt: str | None = None,
        # New fields for the Responses API
        openai_responses_input: list[ResponseInputItemParam] | None = None,
        openai_responses_output: Response | None = None,
    ) -> None:
        request = self.prepare_log_extraction(
            document=document,
            json_schema=json_schema,
            model=model,
            temperature=temperature,
            completion=completion,
            messages=messages,
            openai_messages=openai_messages,
            anthropic_messages=anthropic_messages,
            anthropic_system_prompt=anthropic_system_prompt,
            openai_responses_input=openai_responses_input,
            openai_responses_output=openai_responses_output,
        )
        return self._client._prepared_request(request)


class AsyncExtractions(AsyncAPIResource, BaseExtractionsMixin):
    """Extraction API wrapper for asynchronous usage."""

    async def parse(
        self,
        json_schema: dict[str, Any] | Path | str,
        model: str,
        document: Path | str | IOBase | HttpUrl | None = None,
        documents: list[Path | str | IOBase | HttpUrl] | None = None,
        image_resolution_dpi: int = PydanticUndefined,  # type: ignore[assignment]
        browser_canvas: BrowserCanvas = PydanticUndefined,  # type: ignore[assignment]
        temperature: float = PydanticUndefined,  # type: ignore[assignment]
        modality: Modality = PydanticUndefined,  # type: ignore[assignment]
        reasoning_effort: ChatCompletionReasoningEffort = PydanticUndefined,  # type: ignore[assignment]
        n_consensus: int = PydanticUndefined,  # type: ignore[assignment]
        idempotency_key: str | None = None,
        store: bool = False,
    ) -> RetabParsedChatCompletion:
        """
        Extract structured data from one or more documents asynchronously.

        Args:
            json_schema: JSON schema defining the expected data structure.
            model: The AI model to use.
            document: Single document to process (use either this or documents, not both)
            documents: List of documents to process (use either this or document, not both)
            image_resolution_dpi: Optional image resolution DPI.
            browser_canvas: Optional browser canvas size.
            temperature: Model temperature setting (0-1).
            modality: Modality of the document (e.g., native).
            reasoning_effort: The effort level for the model to reason about the input data.
            n_consensus: Number of consensus extractions to perform (default: 1 which computes a single extraction and the likelihoods comes from the model logprobs)
            idempotency_key: Idempotency key for request
            store: Whether to store the document in the Retab database
        Returns:
            RetabParsedChatCompletion: Parsed response from the API.
        Raises:
            ValueError: If neither document nor documents is provided, or if both are provided
        """
        request = self.prepare_extraction(
            json_schema=json_schema,
            document=document,
            documents=documents,
            image_resolution_dpi=image_resolution_dpi,
            browser_canvas=browser_canvas,
            model=model,
            temperature=temperature,
            modality=modality,
            reasoning_effort=reasoning_effort,
            stream=False,
            n_consensus=n_consensus,
            store=store,
            idempotency_key=idempotency_key,
        )
        response = await self._client._prepared_request(request)
        schema = Schema(json_schema=load_json_schema(json_schema))
        return maybe_parse_to_pydantic(schema, RetabParsedChatCompletion.model_validate(response))

    @as_async_context_manager
    async def stream(
        self,
        json_schema: dict[str, Any] | Path | str,
        model: str,
        document: Path | str | IOBase | HttpUrl | None = None,
        documents: list[Path | str | IOBase | HttpUrl] | None = None,
        image_resolution_dpi: int = PydanticUndefined,  # type: ignore[assignment]
        browser_canvas: BrowserCanvas = PydanticUndefined,  # type: ignore[assignment]
        temperature: float = PydanticUndefined,  # type: ignore[assignment]
        modality: Modality = PydanticUndefined,  # type: ignore[assignment]
        reasoning_effort: ChatCompletionReasoningEffort = PydanticUndefined,  # type: ignore[assignment]
        n_consensus: int = PydanticUndefined,  # type: ignore[assignment]
        idempotency_key: str | None = None,
        store: bool = False,
    ) -> AsyncGenerator[RetabParsedChatCompletion, None]:
        """
        Extract structured data from one or more documents asynchronously with streaming.

        Args:
            json_schema: JSON schema defining the expected data structure.
            model: The AI model to use.
            document: Single document to process (use either this or documents, not both)
            documents: List of documents to process (use either this or document, not both)
            image_resolution_dpi: Optional image resolution DPI.
            browser_canvas: Optional browser canvas size.
            temperature: Model temperature setting (0-1).
            modality: Modality of the document (e.g., native).
            reasoning_effort: The effort level for the model to reason about the input data.
            n_consensus: Number of consensus extractions to perform (default: 1 which computes a single extraction and the likelihoods comes from the model logprobs)
            idempotency_key: Idempotency key for request
            store: Whether to store the document in the Retab database
        Returns:
            AsyncGenerator[RetabParsedChatCompletion, None]: Stream of parsed responses.
        Raises:
            ValueError: If neither document nor documents is provided, or if both are provided

        Usage:
        ```python
        # Single document
        async with retab.documents.extractions.stream(json_schema, model, document=document) as stream:
            async for response in stream:
                print(response)

        # Multiple documents
        async with retab.documents.extractions.stream(json_schema, model, documents=[doc1, doc2]) as stream:
            async for response in stream:
                print(response)
        ```
        """
        request = self.prepare_extraction(
            json_schema=json_schema,
            document=document,
            documents=documents,
            image_resolution_dpi=image_resolution_dpi,
            browser_canvas=browser_canvas,
            model=model,
            temperature=temperature,
            modality=modality,
            reasoning_effort=reasoning_effort,
            stream=True,
            n_consensus=n_consensus,
            store=store,
            idempotency_key=idempotency_key,
        )
        schema = Schema(json_schema=load_json_schema(json_schema))
        ui_parsed_chat_completion_cum_chunk: RetabParsedChatCompletionChunk | None = None
        # Initialize the RetabParsedChatCompletion object
        ui_parsed_completion: RetabParsedChatCompletion = RetabParsedChatCompletion(
            id="",
            created=0,
            model="",
            object="chat.completion",
            likelihoods={},
            choices=[
                RetabParsedChoice(
                    index=0,
                    message=ParsedChatCompletionMessage(content="", role="assistant"),
                    finish_reason=None,
                    logprobs=None,
                )
            ],
        )

        async for chunk_json in self._client._prepared_request_stream(request):
            if not chunk_json:
                continue
            ui_parsed_chat_completion_cum_chunk = RetabParsedChatCompletionChunk.model_validate(chunk_json).chunk_accumulator(ui_parsed_chat_completion_cum_chunk)
            # Basic stuff
            ui_parsed_completion.id = ui_parsed_chat_completion_cum_chunk.id
            ui_parsed_completion.created = ui_parsed_chat_completion_cum_chunk.created
            ui_parsed_completion.model = ui_parsed_chat_completion_cum_chunk.model

            # Update the ui_parsed_completion object
            parsed = unflatten_dict(ui_parsed_chat_completion_cum_chunk.choices[0].delta.flat_parsed)
            likelihoods = unflatten_dict(ui_parsed_chat_completion_cum_chunk.choices[0].delta.flat_likelihoods)
            ui_parsed_completion.choices[0].message.content = json.dumps(parsed)
            ui_parsed_completion.choices[0].message.parsed = parsed
            ui_parsed_completion.likelihoods = likelihoods

            yield maybe_parse_to_pydantic(schema, ui_parsed_completion, allow_partial=True)

        # change the finish_reason to stop
        ui_parsed_completion.choices[0].finish_reason = "stop"
        yield maybe_parse_to_pydantic(schema, ui_parsed_completion)

    async def log(
        self,
        document: Path | str | IOBase | HttpUrl | None,
        json_schema: dict[str, Any],
        model: str,
        temperature: float,
        completion: Any | None = None,
        # The messages can be provided in different formats, we will convert them to the Retab-compatible format
        messages: list[ChatCompletionRetabMessage] | None = None,
        openai_messages: list[ChatCompletionMessageParam] | None = None,
        anthropic_messages: list[MessageParam] | None = None,
        anthropic_system_prompt: str | None = None,
        # New fields for the Responses API
        openai_responses_input: list[ResponseInputItemParam] | None = None,
        openai_responses_output: Response | None = None,
    ) -> None:
        request = self.prepare_log_extraction(
            document=document,
            json_schema=json_schema,
            model=model,
            temperature=temperature,
            completion=completion,
            messages=messages,
            openai_messages=openai_messages,
            anthropic_messages=anthropic_messages,
            anthropic_system_prompt=anthropic_system_prompt,
            openai_responses_input=openai_responses_input,
            openai_responses_output=openai_responses_output,
        )
        return await self._client._prepared_request(request)

File: clients/node/src/resources/documents/extractions.ts
import fs from 'fs';
import path from 'path';
import { SyncAPIResource, AsyncAPIResource } from '../../resource.js';
import { PreparedRequest, FieldUnset } from '../../types/standards.js';
import { 
  DocumentExtractRequestSchema,
  RetabParsedChatCompletion,
  RetabParsedChatCompletionChunk,
  LogExtractionRequest,
  LogExtractionRequestSchema,
  LogExtractionResponse,
  ChatCompletionReasoningEffort
} from '../../types/documents/extractions.js';
import { MIMEData } from '../../types/mime.js';
import { BrowserCanvas } from '../../types/browser_canvas.js';
import { Modality } from '../../types/modalities.js';

// Helper function for MIME document preparation
function prepareMimeDocument(document: string | Buffer | MIMEData): MIMEData {
  if (typeof document === 'string') {
    // Resolve relative paths 
    const resolvedPath = path.resolve(document);
    // If it's a file path, read the file
    if (fs.existsSync(resolvedPath)) {
      const filename = path.basename(resolvedPath);
      const content = fs.readFileSync(resolvedPath);
      const base64Content = content.toString('base64');
      const mimeType = getMimeType(resolvedPath);
      return {
        id: 'doc_' + Date.now(),
        extension: path.extname(filename).slice(1) || 'unknown',
        content: base64Content,
        mime_type: mimeType,
        unique_filename: filename,
        size: content.length,
        filename,
        url: `data:${mimeType};base64,${base64Content}`,
      };
    }
    // Otherwise treat as text content
    const base64Content = Buffer.from(document).toString('base64');
    return {
      id: 'doc_' + Date.now(),
      extension: 'txt',
      content: base64Content,
      mime_type: 'text/plain',
      unique_filename: 'content.txt',
      size: Buffer.byteLength(document),
      filename: 'content.txt',
      url: `data:text/plain;base64,${base64Content}`,
    };
  } else if (Buffer.isBuffer(document)) {
    const base64Content = document.toString('base64');
    return {
      id: 'doc_' + Date.now(),
      extension: 'bin',
      content: base64Content,
      mime_type: 'application/octet-stream',
      unique_filename: 'content.bin',
      size: document.length,
      filename: 'content.bin',
      url: `data:application/octet-stream;base64,${base64Content}`,
    };
  }
  return document as MIMEData;
}

function getMimeType(filePath: string): string {
  const ext = path.extname(filePath).toLowerCase();
  const mimeTypes: Record<string, string> = {
    '.pdf': 'application/pdf',
    '.jpg': 'image/jpeg',
    '.jpeg': 'image/jpeg',
    '.png': 'image/png',
    '.txt': 'text/plain',
    '.json': 'application/json',
    '.html': 'text/html',
    '.xml': 'application/xml',
    '.csv': 'text/csv',
  };
  return mimeTypes[ext] || 'application/octet-stream';
}

function loadJsonSchema(jsonSchema: Record<string, any> | string): Record<string, any> {
  if (typeof jsonSchema === 'string') {
    if (fs.existsSync(jsonSchema)) {
      return JSON.parse(fs.readFileSync(jsonSchema, 'utf-8'));
    }
    return JSON.parse(jsonSchema);
  }
  return jsonSchema;
}

function assertValidModelExtraction(model: string): void {
  // Add model validation logic as needed
  if (!model || typeof model !== 'string') {
    throw new Error('Valid model must be provided for extraction');
  }
}



export class BaseExtractionsMixin {
  public prepareMimeDocument(document: string | Buffer | MIMEData): MIMEData {
    return prepareMimeDocument(document);
  }

  public prepareExtraction(
    jsonSchema: Record<string, any> | string,
    document?: string | Buffer | MIMEData | null,
    documents?: Array<string | Buffer | MIMEData> | null,
    imageResolutionDpi: number | typeof FieldUnset = FieldUnset,
    browserCanvas: BrowserCanvas | typeof FieldUnset = FieldUnset,
    model: string | typeof FieldUnset = FieldUnset,
    temperature: number | typeof FieldUnset = FieldUnset,
    modality: Modality | typeof FieldUnset = FieldUnset,
    reasoningEffort: ChatCompletionReasoningEffort | typeof FieldUnset = FieldUnset,
    stream: boolean = false,
    nConsensus: number | typeof FieldUnset = FieldUnset,
    store: boolean = false,
    idempotencyKey?: string | null
  ): PreparedRequest {
    if (model !== FieldUnset) {
      assertValidModelExtraction(model as string);
    }

    const loadedJsonSchema = loadJsonSchema(jsonSchema);

    // Handle both single document and multiple documents
    if (document !== null && document !== undefined && documents !== null && documents !== undefined) {
      throw new Error('Cannot provide both "document" and "documents" parameters. Use either one.');
    }

    // Convert single document to documents list for consistency
    let processedDocuments: MIMEData[];
    if (document !== null && document !== undefined) {
      processedDocuments = [prepareMimeDocument(document)];
    } else if (documents !== null && documents !== undefined) {
      processedDocuments = documents.map(doc => prepareMimeDocument(doc));
    } else {
      throw new Error('Must provide either "document" or "documents" parameter.');
    }

    // Build request dictionary with only provided fields
    const requestDict: any = {
      json_schema: loadedJsonSchema,
      documents: processedDocuments,
      stream,
      store,
    };

    if (model !== FieldUnset) requestDict.model = model;
    if (temperature !== FieldUnset) requestDict.temperature = temperature;
    if (modality !== FieldUnset) requestDict.modality = modality;
    if (reasoningEffort !== FieldUnset) requestDict.reasoning_effort = reasoningEffort;
    if (nConsensus !== FieldUnset) requestDict.n_consensus = nConsensus;
    if (imageResolutionDpi !== FieldUnset) requestDict.image_resolution_dpi = imageResolutionDpi;
    if (browserCanvas !== FieldUnset) requestDict.browser_canvas = browserCanvas;

    // Validate DocumentExtractRequest data
    const request = DocumentExtractRequestSchema.parse(requestDict);

    return {
      method: 'POST',
      url: '/v1/documents/extractions',
      data: request,
      idempotencyKey,
    };
  }

  public prepareLogExtraction(logRequest: LogExtractionRequest): PreparedRequest {
    const request = LogExtractionRequestSchema.parse(logRequest);
    return {
      method: 'POST',
      url: '/v1/documents/log_extraction',
      data: request,
    };
  }
}

export class Extractions extends SyncAPIResource {
  private mixin = new BaseExtractionsMixin();

  async extract(
    jsonSchema: Record<string, any> | string,
    options: {
      document?: string | Buffer | MIMEData;
      documents?: Array<string | Buffer | MIMEData>;
      imageResolutionDpi?: number;
      browserCanvas?: BrowserCanvas;
      model?: string;
      temperature?: number;
      modality?: Modality;
      reasoningEffort?: ChatCompletionReasoningEffort;
      stream?: boolean;
      nConsensus?: number;
      store?: boolean;
      idempotencyKey?: string;
    } = {}
  ): Promise<RetabParsedChatCompletion> {
    const preparedRequest = this.mixin.prepareExtraction(
      jsonSchema,
      options.document,
      options.documents,
      options.imageResolutionDpi || FieldUnset,
      options.browserCanvas || FieldUnset,
      options.model || FieldUnset,
      options.temperature || FieldUnset,
      options.modality || FieldUnset,
      options.reasoningEffort || FieldUnset,
      options.stream || false,
      options.nConsensus || FieldUnset,
      options.store || false,
      options.idempotencyKey
    );

    const response = await this._client._preparedRequest(preparedRequest);
    return response as RetabParsedChatCompletion;
  }

  async *extractStream(
    jsonSchema: Record<string, any> | string,
    options: {
      document?: string | Buffer | MIMEData;
      documents?: Array<string | Buffer | MIMEData>;
      imageResolutionDpi?: number;
      browserCanvas?: BrowserCanvas;
      model?: string;
      temperature?: number;
      modality?: Modality;
      reasoningEffort?: ChatCompletionReasoningEffort;
      nConsensus?: number;
      store?: boolean;
      idempotencyKey?: string;
    } = {}
  ): AsyncGenerator<RetabParsedChatCompletionChunk, void, unknown> {
    const preparedRequest = this.mixin.prepareExtraction(
      jsonSchema,
      options.document,
      options.documents,
      options.imageResolutionDpi || FieldUnset,
      options.browserCanvas || FieldUnset,
      options.model || FieldUnset,
      options.temperature || FieldUnset,
      options.modality || FieldUnset,
      options.reasoningEffort || FieldUnset,
      true, // stream = true
      options.nConsensus || FieldUnset,
      options.store || false,
      options.idempotencyKey
    );

    const streamGenerator = await this._client._preparedRequestStream(preparedRequest);
    for await (const chunk of streamGenerator) {
      yield chunk as RetabParsedChatCompletionChunk;
    }
  }

  async logExtraction(logRequest: LogExtractionRequest): Promise<LogExtractionResponse> {
    const preparedRequest = this.mixin.prepareLogExtraction(logRequest);
    const response = await this._client._preparedRequest(preparedRequest);
    return response as LogExtractionResponse;
  }
}

export class AsyncExtractions extends AsyncAPIResource {
  private mixin = new BaseExtractionsMixin();

  async extract(
    jsonSchema: Record<string, any> | string,
    options: {
      document?: string | Buffer | MIMEData;
      documents?: Array<string | Buffer | MIMEData>;
      imageResolutionDpi?: number;
      browserCanvas?: BrowserCanvas;
      model?: string;
      temperature?: number;
      modality?: Modality;
      reasoningEffort?: ChatCompletionReasoningEffort;
      stream?: boolean;
      nConsensus?: number;
      store?: boolean;
      idempotencyKey?: string;
    } = {}
  ): Promise<RetabParsedChatCompletion> {
    const preparedRequest = this.mixin.prepareExtraction(
      jsonSchema,
      options.document,
      options.documents,
      options.imageResolutionDpi || FieldUnset,
      options.browserCanvas || FieldUnset,
      options.model || FieldUnset,
      options.temperature || FieldUnset,
      options.modality || FieldUnset,
      options.reasoningEffort || FieldUnset,
      options.stream || false,
      options.nConsensus || FieldUnset,
      options.store || false,
      options.idempotencyKey
    );

    const response = await this._client._preparedRequest(preparedRequest);
    return response as RetabParsedChatCompletion;
  }

  async *extractStream(
    jsonSchema: Record<string, any> | string,
    options: {
      document?: string | Buffer | MIMEData;
      documents?: Array<string | Buffer | MIMEData>;
      imageResolutionDpi?: number;
      browserCanvas?: BrowserCanvas;
      model?: string;
      temperature?: number;
      modality?: Modality;
      reasoningEffort?: ChatCompletionReasoningEffort;
      nConsensus?: number;
      store?: boolean;
      idempotencyKey?: string;
    } = {}
  ): AsyncGenerator<RetabParsedChatCompletionChunk, void, unknown> {
    const preparedRequest = this.mixin.prepareExtraction(
      jsonSchema,
      options.document,
      options.documents,
      options.imageResolutionDpi || FieldUnset,
      options.browserCanvas || FieldUnset,
      options.model || FieldUnset,
      options.temperature || FieldUnset,
      options.modality || FieldUnset,
      options.reasoningEffort || FieldUnset,
      true, // stream = true
      options.nConsensus || FieldUnset,
      options.store || false,
      options.idempotencyKey
    );

    const streamGenerator = this._client._preparedRequestStream(preparedRequest);
    for await (const chunk of streamGenerator) {
      yield chunk as RetabParsedChatCompletionChunk;
    }
  }

  async logExtraction(logRequest: LogExtractionRequest): Promise<LogExtractionResponse> {
    const preparedRequest = this.mixin.prepareLogExtraction(logRequest);
    const response = await this._client._preparedRequest(preparedRequest);
    return response as LogExtractionResponse;
  }
}